Add support Slide cover

ualex73 opened the pull request:
## Breaking Change:

None

## Description:

Add support Slide cover integration (Cloud API).

**Related issue (if applicable):** None

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) (if applicable):** home-assistant/home-assistant.io#10127

## Example entry for `configuration.yaml` (if applicable):
```yaml
slide:
  username: YOUR_SLIDE_APP_USERNAME
  password: YOUR_SLIDE_APP_PASSWORD
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [X] There is no commented out code in this PR.
  - [X] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [X] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [X] Untested files have been added to `.coveragerc`.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

ualex73 committed:
Add support GoSlide cover

ualex73 committed:
Fixed Parameters differ from overridden
Fixed Removed other pylint warnings

ualex73 committed:
Renamed GoSlide to Slide, because of Innovation in Motion rebranding

ualex73 committed:
Fixed codeowners file

ualex73 committed:
Fixed requirements file

ualex73 commented:
Will anybody from the HA team will look into this PR?

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,159 @@
+async def async_setup(hass, config):
+    """Set up the Slide platform."""
+
+    # pylint: disable=unused-argument
```
This is already globally disabled.
```suggestion
```

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,159 @@
+            hass.data[DOMAIN][SLIDES][uid] = slidenew
+            _LOGGER.debug("Updated entry=%s", slidenew)
+
+    if DOMAIN not in config:
```
This can't happen.

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,159 @@
+            if "pos" in slide["device_info"]:
+                slidenew["online"] = True
+                slidenew["pos"] = slide["device_info"]["pos"]
+                if slidenew["pos"] < 0:
```
```suggestion
                slidenew["pos"] = max(0, min(1, slidenew["pos"]))
```

ualex73 replied:
Made the change, and learned from it.

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,159 @@
+    timeout = config[DOMAIN][CONF_TIMEOUT]
+    retry = config[DOMAIN][CONF_RETRY]
+
+    from goslideapi import GoSlideCloud
```
3rd party imports can be moved to the top of the module.

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,159 @@
+    # pylint: disable=broad-except
+    try:
+        result = await hass.data[DOMAIN][API].login()
+    except Exception as err:
```
Please raise library specific exceptions in the library. We should only catch expected errors.

ualex73 replied:
Not done yet, because i need to make a few modifications to the library (i will do that tomorrow)

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,159 @@
+        result = await hass.data[DOMAIN][API].login()
+    except Exception as err:
+        _LOGGER.error("Error connecting to Slide Cloud: %s", str(err))
+        _LOGGER.error("Going to retry in %s seconds", retry)
```
Please combine the two logging messages into one message.

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,159 @@
+    except Exception as err:
+        _LOGGER.error("Error connecting to Slide Cloud: %s", str(err))
+        _LOGGER.error("Going to retry in %s seconds", retry)
+        async_call_later(hass, retry, async_setup(hass, config))
```
If the credentials are incorrect there's no point in retrying later.

ualex73 replied:
Invalid credentials can't happen in this exception, this is captured in the goslideapi. The login returns a "False" when this happens. The exception captures mainly connection issues, that's why i added a retry here.

MartinHjelmare replied:
Please add exceptions for invalid credentials too.

ualex73 replied:
Ok, i will have a look into this, currently it doesn't raise an exception in the goslide-api library.

ualex73 replied:
Invalid authentication will raise an exception now (it will not retry).

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,159 @@
+        async_call_later(hass, retry, async_setup(hass, config))
+        return True
+
+    # pylint: disable=no-else-return
```
Please don't disable this.
```suggestion
```

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,159 @@
+    if result is None:
+        _LOGGER.error("Slide API returned unknown error during " "authentication")
+        return False
+    elif result is False:
```
```suggestion
    if result is False:
```
It's better to raise specific exceptions on different types of error.

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,159 @@
+                    CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL
+                ): cv.time_period,
+                vol.Optional(CONF_TIMEOUT, default=DEFAULT_TIMEOUT): cv.positive_int,
+                vol.Optional(CONF_RETRY, default=DEFAULT_RETRY): cv.positive_int,
```
Why do we need a config option for retry?

ualex73 replied:
The retry is used when an exception is caught during setup, thus the seconds between we should retry the setup of the 'slide' component again. The exception are normally connection errors, so not authentication issues. When authentication fails, the setup will stop and will not retry.

MartinHjelmare replied:
And why does it need to be configurable?

ualex73 replied:
That's fine with me too, i will remove this in the next commit.

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,159 @@
+                vol.Optional(
+                    CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL
+                ): cv.time_period,
+                vol.Optional(CONF_TIMEOUT, default=DEFAULT_TIMEOUT): cv.positive_int,
```
Why do we need a config option for timeout?

ualex73 replied:
The timeout option is the aiohttp timeout for the cloud requests. The default is 300+ seconds, that's why i make it configurable (default is 30 seconds).

MartinHjelmare replied:
Isn't it enough to always use th default?

ualex73 replied:
That's fine with me too, i will remove this in the next commit.

ualex73 replied:
Made the change, so removed.

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,159 @@
+
+    from goslideapi import GoSlideCloud
+
+    hass.data[DOMAIN][API] = GoSlideCloud(username, password, timeout)
```
Please have `GoSlideCloud` accept an optional aiohttp session. We want all integration that use aiohttp client to use the home assistant aiohttp session. Use the aiohttp client helper to get the session instance.

ualex73 replied:
At this moment the goslideapi doesn't use the aiohttp session, it uses the simple "aiohttp.request" version. I will have a look if i can change/improve this in the next version.

ualex73 replied:
I will resolve this, because I don't use an aiohttp.session at this moment.

MartinHjelmare commented on homeassistant/components/slide/const.py:
```diff
@@ -0,0 +1,10 @@
+COMPONENT = "cover"
+DOMAIN = "slide"
+SLIDES = "slides"
+STATE_OFFLINE = "offline"
```
This isn't used.

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+# pylint: disable=unused-argument
```
```suggestion
```

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+# pylint: disable=unused-argument
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up cover(s) for Go Slide platform."""
+    entities = []
```
Add a guard clause that checks if discovery_info is None and return if so.

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+    """Set up cover(s) for Go Slide platform."""
+    entities = []
+
+    for key in hass.data[DOMAIN][SLIDES]:
```
```suggestion
    for slide in hass.data[DOMAIN][SLIDES].values():
```

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+    entities = []
+
+    for key in hass.data[DOMAIN][SLIDES]:
+        _LOGGER.debug("Setting up Slide entity: %s", hass.data[DOMAIN][SLIDES][key])
```
```suggestion
        _LOGGER.debug("Setting up Slide entity: %s", slide)
```

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+
+    for key in hass.data[DOMAIN][SLIDES]:
+        _LOGGER.debug("Setting up Slide entity: %s", hass.data[DOMAIN][SLIDES][key])
+        entities.append(SlideCover(hass, hass.data[DOMAIN][SLIDES][key]))
```
```suggestion
        entities.append(SlideCover(slide))
```

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+class SlideCover(CoverDevice):
+    """Representation of a Go Slide cover."""
+
+    def __init__(self, hass, slide):
```
Please pass in the api dict instead.
```suggestion
    def __init__(self, api, slide):
```

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+
+    def __init__(self, hass, slide):
+        """Initialize the cover."""
+        self._hass = hass
```
```suggestion
self._api = api
self._slide = slide
```

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+        self._mac = slide["mac"]
+        self._id = slide["id"]
+        self._name = slide["name"]
+        self._entity_id = ENTITY_ID_FORMAT.format(slugify("slide_" + self._mac))
```
We can use the mac address as a unique id. Then the user can change entity_id and name via GUI. Please implement the `unique_id` entity property:
https://developers.home-assistant.io/docs/en/entity_index.html#generic-properties
https://developers.home-assistant.io/docs/en/entity_registry_index.html#unique-id-requirements
```suggestion
```

ualex73 replied:
Made the change, and it is a good improvement - thanks 👍

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+        self._is_closed = None
+
+    @property
+    def entity_id(self):
```
Remove this. The name property will be used as default template for entity_id.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+        return self._name
+
+    @property
+    def state(self):
```
Don't overwrite the state property. It's handled by the base cover entity class.

ualex73 replied:
Made the change, i added properties for is_closing/is_opening and added a value for opening/closing/opened/closed in the __init__.py.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+        return DEVICE_CLASS_CURTAIN
+
+    @property
+    def supported_features(self):
```
We don't need to overwrite this. It will be set by the base cover entity class according to what properties we implement.

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,159 @@
+        else:
+            _LOGGER.warning("Slide API returned 0 slides")
+
+        for key in hass.data[DOMAIN][SLIDES]:
```
```suggestion
        for slide in hass.data[DOMAIN][SLIDES].values():
```

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,159 @@
+            _LOGGER.warning("Slide API returned 0 slides")
+
+        for key in hass.data[DOMAIN][SLIDES]:
+            hass.data[DOMAIN][SLIDES][key]["pos"] = None
```
```suggestion
            slide["pos"] = None
```

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,159 @@
+                    slide["device_info"],
+                )
+
+            hass.data[DOMAIN][SLIDES][uid] = slidenew
```
If we update the existing dict, and make sure we set a default, we can simplify downstream code.
```suggestion
            old_slide = hass.data[DOMAIN][SLIDES].setdefault(uid, {})
            old_slide.update(slidenew)
```

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+    @property
+    def current_cover_position(self):
+        """Return the current position of cover shutter."""
+        if self._hass.data[DOMAIN][SLIDES][self._mac]["pos"] is None:
```
```suggestion
        pos = self._slide["pos"]
        if pos is not None:
            pos = int(pos * 100)
        return pos
```

ualex73 replied:
This one is also noted below, so i will put it on resolved.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+    async def async_open_cover(self, **kwargs):
+        """Open the cover."""
+        self._is_closed = False
+        await self._hass.data[DOMAIN][API].slideopen(self._id)
```
```suggestion
        await self._api.slideopen(self._id)
```

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+    async def async_close_cover(self, **kwargs):
+        """Close the cover."""
+        self._is_closed = True
+        await self._hass.data[DOMAIN][API].slideclose(self._id)
```
```suggestion
        await self._api.slideclose(self._id)
```

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+
+    async def async_stop_cover(self, **kwargs):
+        """Stop the cover."""
+        await self._hass.data[DOMAIN][API].slidestop(self._id)
```
```suggestion
        await self._api.slidestop(self._id)
```

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+        """Move the cover to a specific position."""
+        position = kwargs[ATTR_POSITION] / 100
+
+        if self._hass.data[DOMAIN][SLIDES][self._mac]["pos"] is not None:
```
```suggestion
        pos = self._slide["pos"]
        if pos is not None:
```

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+
+        if self._hass.data[DOMAIN][SLIDES][self._mac]["pos"] is not None:
+            self._is_closed = (
+                position > self._hass.data[DOMAIN][SLIDES][self._mac]["pos"]
```
```suggestion
                position > pos
```

ualex73 replied:
Made the change.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,128 @@
+                position > self._hass.data[DOMAIN][SLIDES][self._mac]["pos"]
+            )
+
+        await self._hass.data[DOMAIN][API].slidesetposition(self._id, position)
```
```suggestion
        await self._api.slidesetposition(self._id, position)
```

It's a lot easier to read the method names of the library if we use snake_case names, which is PEP8 recommended. I suggest to update the library and name the methods etc with snake_case names, eg `slide_set_position`.

ualex73 replied:
I need to rework the library, i will do that tomorrow.

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,159 @@
+
+            # Check if we have 'pos' (OK) or 'code' (NOK)
+            if "pos" in slide["device_info"]:
+                slidenew["online"] = True
```
We never use the online key.

ualex73 replied:
I added the property available, so it is used now.

ualex73 commented:
@MartinHjelmare I am reworking the __init__.py/cover.py/const.py at this moment and i will test the new code, before committing it again. I have replied to a few code reviews (the exception and login, retry and timeout configuration items)

ualex73 committed:
Removed pylint: disable=unused-argument
Removed DOMAIN not exist check
Changed if to min/max
Changed 3rd party import to top of the module
Removed timeout/retry parameters
Removed unused constants
Added check for discovery_info is none
Changed pass slide object instead of full hass object
Changed pass api object instead of full hass object
Added unique_id functionality
Removed entity_id/name properties
Removed supported_features/state functions

ualex73 committed:
Fixed unused variables

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,121 @@
+    @property
+    def device_state_attributes(self):
+        """Return device specific state attributes."""
+        return {ATTR_ID: self._id, ATTR_FRIENDLY_NAME: self._friendly_name}
```
Please implement the `name` property instead of setting the friendly name state attribute.

MartinHjelmare commented on homeassistant/components/slide/cover.py:
```diff
@@ -0,0 +1,121 @@
+    @property
+    def current_cover_position(self):
+        """Return the current position of cover shutter."""
+        if self._slide["pos"] is None:
```
This is cleaner:
```py
pos = self._slide["pos"]
if pos is not None:
    pos = int(pos * 100)
return pos
```

MartinHjelmare commented on homeassistant/components/slide/manifest.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "domain": "slide",
```
`slide` is too generic I think. Can we name it `goslide` instead? Our standard is to name integrations after the brand names.

ualex73 replied:
They are using 'slide' as a brand/product name (https://slide.store). It was initially goslide, but have rebranded everything now to slide. In theory this matches the guideline, but if you prefer to be named 'goslide', i will rebrand again.

MartinHjelmare replied:
@balloob what do you think?

balloob replied:
Slide is fine. I doubt that we'll ever get smart slides integrated

ualex73 committed:
Changed Slide API uses snake names
Changed Improved exception handling
Changed Updated Slide API to 0.50.0

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -133,20 +132,19 @@
             str(err),
             DEFAULT_RETRY,
         )
         async_call_later(hass, DEFAULT_RETRY, async_setup(hass, config))
```
Use a lambda like here:
https://github.com/home-assistant/home-assistant/blob/ed5d3dba0e259a9648da13093a72769a2e7874c0/homeassistant/components/influxdb/__init__.py#L175

The target function for the event helper should only accept one argument.

ualex73 replied:
I tried the lambda example, but it won't work - because the async_setup is 'async' and it needs to be awaited?

MartinHjelmare replied:
Right, then create a coroutine function that accepts the event time argument and awaits `async_setup`.

ualex73 replied:
I think I need to put it into a separate class to do this.

Thanks for you patience with me and help, as you notice, I am not a Python expert.

ualex73 replied:
Ok, found a quicker solution 👍 

```python
...
    async def retry_setup(now):
        """Retry setup if a connection/timeout happens on Slide API."""
        await async_setup(hass, config)
...
...
        async_call_later(hass, DEFAULT_RETRY, retry_setup)
        return True
...
```

MartinHjelmare replied:
Yeah, that was what I was thinking.

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,156 @@
+
+    hass.data[DOMAIN][API] = GoSlideCloud(username, password)
+
+    # pylint: disable=broad-except
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,156 @@
+    # pylint: disable=broad-except
+    try:
+        result = await hass.data[DOMAIN][API].login()
+    except (ClientConnectionError, ClientConnectorError, TimeoutError) as err:
```
It's better if the library raises library specific exceptions, with a common library specific parent exception.

ualex73 committed:
Changed moved exceptions into goslide-api
Changed retry setup into coroutine

ualex73 commented:
@MartinHjelmare I am still new with the process, but it looks like I don't have any open points/changed pending from my side?

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,157 @@
+    except (goslideapi.ClientConnectionError, goslideapi.ClientTimeoutError) as err:
+        _LOGGER.error(
+            "Error connecting to Slide Cloud: %s, going to retry in %s seconds",
+            str(err),
```
The exception argument has a `__str__` magic method.
```suggestion
            err,
```

MartinHjelmare commented on homeassistant/components/slide/__init__.py:
```diff
@@ -0,0 +1,157 @@
+        async_call_later(hass, DEFAULT_RETRY, retry_setup)
+        return True
+
+    if result:
```
If we invert this check and log error and return False if true, we can outdent the debug log.
```suggestion
    if not result:
```

MartinHjelmare reviewed (commented):
It's not required for this PR, but it would be good to update the library to use aiohttp client session api.

ualex73 committed:
Changed str(err) to err
Changed invert if result to if not result

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare commented:
Can be merged when build passes.
